trie/triedb/pathdb/history.go

```diff
@@ -0,0 +1,594 @@
+			return nil, nil, errors.New("storage index buffer is gapped")
+		}
+		if uint32(len(r.storageIndexes)) < end {
+			return nil, nil, errors.New("storage index buffer is corrupted")
+		}
+		index.decode(r.storageIndexes[start:end])
+
+		if last != (common.Hash{}) {
+			if bytes.Compare(last.Bytes(), index.Hash.Bytes()) >= 0 {
+				return nil, nil, errors.New("storage slot is not in order")
+			}
+		}
+		if index.Offset != r.lastStorageRead {
+			return nil, nil, errors.New("storage data buffer is gapped")
+		}
+		sEnd := index.Offset + uint32(index.Length)
+		if uint32(len(r.storageData)) < sEnd {
+			return nil, nil, errors.New("storage data buffer is corrupted")
+		}
+		storage[index.Hash] = r.storageData[r.lastStorageRead:sEnd]
+		list = append(list, index.Hash)
+
+		last = index.Hash
+		r.lastSIndexRead = end
+		r.lastStorageRead = sEnd
+	}
+	return list, storage, nil
+}
+
+// decode deserializes the account and storage data from the provided byte stream.
+func (h *history) decode(accountData, storageData, accountIndexes, storageIndexes []byte) error {
+	var (
+		accounts    = make(map[common.Hash][]byte)
+		storages    = make(map[common.Hash]map[common.Hash][]byte)
+		accountList []common.Hash
+		storageList = make(map[common.Hash][]common.Hash)
+
+		r = &decoder{
+			accountData:    accountData,
+			storageData:    storageData,
+			accountIndexes: accountIndexes,
+			storageIndexes: storageIndexes,
+		}
+	)
+	if err := r.verify(); err != nil {
+		return err
+	}
+	for i := 0; i < len(accountIndexes)/accountIndexSize; i++ {
+		// Resolve account first
+		aIndex, aData, err := r.readAccount(i)
+		if err != nil {
+			return err
+		}
+		accounts[aIndex.Hash] = aData
+		accountList = append(accountList, aIndex.Hash)
+
+		// Resolve storage slots
+		slotList, slotData, err := r.readStorage(aIndex)
+		if err != nil {
+			return err
+		}
+		if len(slotList) > 0 {
+			storageList[aIndex.Hash] = slotList
+			storages[aIndex.Hash] = slotData
+		}
+	}
+	h.accounts = accounts
+	h.accountList = accountList
+	h.storages = storages
+	h.storageList = storageList
+	return nil
+}
+
+// readHistory reads and decodes the state history object by the given id.
+func readHistory(freezer *rawdb.ResettableFreezer, id uint64) (*history, error) {
+	blob := rawdb.ReadStateHistoryMeta(freezer, id)
+	if len(blob) == 0 {
+		return nil, fmt.Errorf("state history not found %d", id)
+	}
+	var m meta
+	if err := m.decode(blob); err != nil {
+		return nil, err
+	}
+	var (
+		dec            = history{meta: &m}
+		accountData    = rawdb.ReadStateAccountHistory(freezer, id)
+		storageData    = rawdb.ReadStateStorageHistory(freezer, id)
+		accountIndexes = rawdb.ReadStateAccountIndex(freezer, id)
+		storageIndexes = rawdb.ReadStateStorageIndex(freezer, id)
+	)
+	if err := dec.decode(accountData, storageData, accountIndexes, storageIndexes); err != nil {
+		return nil, err
+	}
+	return &dec, nil
+}
+
+// writeStateHistory writes the state history with provided state set. After
```

Comment:
It's a bit weird that we have "readHistory" but "writeStateHistory". Any specific reason that the two are not symmetrically named? If not, perhaps we can rename one or the other?